### PR TITLE
Use our own either type

### PR DESCRIPTION
**Pull request by taiki-e**
Closes https://github.com/rust-lang-nursery/futures-rs/issues/1152

### CODE REVIEW

File: futures-util/src/future/either.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+/// Combines two different futures, streams, or sinks having the same associated types into a single
+/// type.
+#[derive(Debug, Clone)]
+pub enum Either<A, B> {
+    /// First branch of the type
+    Left(A),
+    /// Second branch of the type
+    Right(B),
+}
+
```
Comment (cramertj) [inline futures-util/src/future/either.rs]: Would you mind also adding an `into_inner` for `Either<T, T>`?

File: futures-util/src/future/mod.rs [hunk 2]
```diff
@@ -186,60 +187,19 @@ pub trait FutureExt: Future {
     /// wrapped version of them.
     ///
     /// Also note that if both this and the second future have the same
-    /// success/error type you can use the `Either::split` method to
+    /// output type you can use the `Either::factor_first` method to
     /// conveniently extract out the value at the end.
-    ///
-    /// # Examples
```
Comment (cramertj) [inline futures-util/src/future/mod.rs]: Why remove these doc examples?

**Comment by cramertj** [comment]
works for me, modulo the removal of doc examples and the introduction of `into_inner`

File: futures-util/src/future/mod.rs [hunk 3]
```diff
@@ -191,40 +192,35 @@ pub trait FutureExt: Future {
...
+    ///             Either::Left((x, b)) => b.map(move |y| (x, y)).left_future(),
+    ///             Either::Right((y, a)) => a.map(move |x| (x, y)).right_future(),
     ///         }
-    ///     }))
+    ///     })
     /// }
     /// ```
-    fn select<B>(self, other: B) -> Select<Self, B::Future>
-        where B: IntoFuture, Self: Sized
+    fn select<B>(self, other: B) -> Select<Self, B>
```
Comment (cramertj) [inline futures-util/src/future/mod.rs]: We just changed the `.join` methods into free functions-- seems like that would make sense for `select` as well?

Comment (taiki-e) [reply futures-util/src/future/mod.rs]: Thanks for pointing that out, it makes sense. And I remembered that I forgot to write PR for changing `StreamExt::select` to a function 😅 .

**Comment by cramertj** [comment]
Nice! thanks for this :)

**Pull request merged**